### PR TITLE
Sparse index: integrate with `read-tree`

### PR DESCRIPTION
**Pull request by vdye**
This pull request is more-or-less split into "fixups" and "integrating sparse index with `read-tree`".

## Fixups
- Fix syntax in some of the `diff` tests (I can't seem to get `echo` to work in the tests, so it's replaced with a custom `edit-contents`)
- Fix `git status` reporting differences nested in sparse directory
  - In the test relevant to this commit, `git status` was previously reporting changes nested inside of a sparse directory as the whole directory being "modified" (but _only_ in the sparse index). Fixed by setting the `recursive` flag when performing the diff
- Modify the way `prime_cache_tree_rec` identifies sparse directories when rebuilding the cache tree
  - The previous implementation was breaking when `read-tree` was sparse-enabled in the `t1091` tests. The prior checks used weren't correctly identifying sparse cache entries in the case that the sparse patterns were broken.
- Change the test looking for sparse index expand and collapse to use `git mv` (which, as far as I can tell, isn't slated for integration with sparse index in the immediate future)
  - This one is a `fixup???` since the commit message for the original would need to be changed, since it calls out `read-tree` by name

## `read-tree`
Other than the standard changes to enable sparse index:
- Expand the index when `--no-sparse-checkout` is provided
  - The option effectively removes the `skip-worktree` flag from all entries, which eventually leads to the command attempting to checkout a sparse directory
- Expand the index when the prefix is equivalent to or is nested inside a sparse directory
  -  When `--prefix` is specified, it forces the `traverse_path` in `unpack_trees`/`traverse_trees` to start at the given prefix. However, the operation of those functions implicitly assumes (requires?) that the traverse path is something it has actually traversed into (which it can't be for a sparse directory). So, if the prefix is a sparse directory (e.g., `folder1` in the `t1092` tests) or is nested in one (e.g., `folder2/0`), the prefix is expanded before continuing with `unpack_trees`

## Performance testing
```
Test                                                   ms/vfs-2.33.0     HEAD                  
-----------------------------------------------------------------------------------------------
2000.2: git status (full-v3)                           0.74(0.28+0.37)   0.97(0.35+0.44) +31.1%
2000.3: git status (full-v4)                           0.80(0.30+0.39)   0.92(0.46+0.40) +15.0%
2000.4: git status (sparse-v3)                         0.56(0.11+0.69)   0.50(0.10+0.71) -10.7%
2000.5: git status (sparse-v4)                         0.54(0.12+0.80)   0.51(0.11+0.72) -5.6% 
2000.22: git reset (full-v3)                           0.78(0.35+0.36)   0.77(0.35+0.27) -1.3% 
2000.23: git reset (full-v4)                           0.84(0.47+0.25)   0.90(0.49+0.29) +7.1% 
2000.24: git reset (sparse-v3)                         0.36(0.07+0.77)   0.46(0.09+1.09) +27.8%
2000.25: git reset (sparse-v4)                         0.40(0.09+0.65)   0.44(0.09+0.63) +10.0%
2000.26: git reset --hard (full-v3)                    1.03(0.61+0.39)   1.11(0.62+0.43) +7.8% 
2000.27: git reset --hard (full-v4)                    1.12(0.65+0.40)   1.18(0.70+0.41) +5.4% 
2000.28: git reset --hard (sparse-v3)                  0.51(0.08+0.31)   0.63(0.10+0.39) +23.5%
2000.29: git reset --hard (sparse-v4)                  0.49(0.08+0.30)   0.56(0.09+0.33) +14.3%
2000.30: git read-tree -mu HEAD (full-v3)              1.29(0.96+0.30)   1.45(1.12+0.34) +12.4%
2000.31: git read-tree -mu HEAD (full-v4)              1.28(0.96+0.30)   1.41(1.10+0.33) +10.2%
2000.32: git read-tree -mu HEAD (sparse-v3)            1.66(1.25+0.40)   0.13(0.07+0.05) -92.2%
2000.33: git read-tree -mu HEAD (sparse-v4)            1.61(1.22+0.38)   0.13(0.07+0.05) -91.9%
```

### CODE REVIEW

**Comment by derrickstolee** [comment]
> ## Fixups
> * Fix syntax in some of the `diff` tests (I can't seem to get `echo` to work in the tests, so it's replaced with a custom `edit-contents`)

The redirection character doesn't work as it encompasses the entire `test_all_match` function call. Unfortunately the `edit-contents` script is our best approach right now.

> * Fix `git status` reporting differences nested in sparse directory
>   
>   * In the test relevant to this commit, `git status` was previously reporting changes nested inside of a sparse directory as the whole directory being "modified" (but _only_ in the sparse index). Fixed by setting the `recursive` flag when performing the diff

Wow! Good catch.

> * Modify the way `prime_cache_tree_rec` identifies sparse directories when rebuilding the cache tree
>   
>   * The previous implementation was breaking when `read-tree` was sparse-enabled in the `t1091` tests. The prior checks used weren't correctly identifying sparse cache entries in the case that the sparse patterns were broken.

I'll go read the commit for this, but it sounds like we can get into strange situations where the sparse-checkout patterns change to no longer be cone-mode, but we still have a sparse index. That's a good case to handle.

> * Change the test looking for sparse index expand and collapse to use `git mv` (which, as far as I can tell, isn't slated for integration with sparse index in the immediate future)

Good idea. It's not high on the list. We might eventually drop the "definitely expand" test, but for now we can continue chasing commands that still expand it.

>   * This one is a `fixup???` since the commit message for the original would need to be changed, since it calls out `read-tree` by name

Right. We could squash it in when upstreaming that commit. You could also do a `git commit --fixup <oid>` followed by a `git commit --fixup=reword:<oid>` to modify the commit message and we will naturally squash them together with the new message in the 2.34 rebase.

> ## `read-tree`
> Other than the standard changes to enable sparse index:
> 
> * Expand the index when `--no-sparse-checkout` is provided
>   
>   * The option effectively removes the `skip-worktree` flag from all entries, which eventually leads to the command attempting to checkout a sparse directory

Makes sense.

> * Expand the index when the prefix is equivalent to or is nested inside a sparse directory
>   
>   * When `--prefix` is specified, it forces the `traverse_path` in `unpack_trees`/`traverse_trees` to start at the given prefix. However, the operation of those functions implicitly assumes (requires?) that the traverse path is something it has actually traversed into (which it can't be for a sparse directory). So, if the prefix is a sparse directory (e.g., `folder1` in the `t1092` tests) or is nested in one (e.g., `folder2/0`), the prefix is expanded before continuing with `unpack_trees`

This also makes sense for safety. We could extend this later to refuse to call `unpack_trees()` if the prefix is within a sparse directory (or generally outside of the sparse-checkout definition). But, I doubt that will actually be worth the effort since users shouldn't be specifying the prefix this way.

> ## Performance testing
> ```
> Test                                                   ms/vfs-2.33.0     HEAD                  
> -----------------------------------------------------------------------------------------------
> 2000.30: git read-tree -mu HEAD (full-v3)              1.29(0.96+0.30)   1.45(1.12+0.34) +12.4%
> 2000.31: git read-tree -mu HEAD (full-v4)              1.28(0.96+0.30)   1.41(1.10+0.33) +10.2%
> 2000.32: git read-tree -mu HEAD (sparse-v3)            1.66(1.25+0.40)   0.13(0.07+0.05) -92.2%
> 2000.33: git read-tree -mu HEAD (sparse-v4)            1.61(1.22+0.38)   0.13(0.07+0.05) -91.9%
> ```

Feel free to delete the unnecessary rows (like this) when talking about them in your commit messages.

Now... to go looking at the code.

File: read-cache.c
Comment (derrickstolee) [inline read-cache.c:617]: This implementation _could_ be `return index_name_stage_pos() >= 0;` but we specifically don't want to expand the index in the case of a sparse directory. Perhaps it would be better to extend `index_name_stage_pos()` to have an extra parameter that controls that expansion. Then `index_entry_exists()` could still use that same binary search.

It would be worth commenting on the declaration of `index_entry_exists()` that it does not expand a sparse index and hence will return 0 for paths within sparse directory entries.

File: wt-status.c
Comment (derrickstolee) [inline wt-status.c]: I wonder if there is any negative to setting `recursive = 1` in all cases. My guess is that it will be correct, but possibly slower? It would be great if it wasn't actually any slower and we don't need the special case.

File: t/t1092-sparse-checkout-compatibility.sh
Comment (derrickstolee) [inline t/t1092-sparse-checkout-compatibility.sh]: Great job finding and fixing this bug!

File: builtin/read-tree.c
Comment (derrickstolee) [inline builtin/read-tree.c]: A very good approach so far. These are reasonable protections.

File: t/t1092-sparse-checkout-compatibility.sh
Comment (derrickstolee) [inline t/t1092-sparse-checkout-compatibility.sh]: One case that might be interesting is a situation where we do a two- or three-way merge and two of the refs changed files such as `folder1/a` and `folder1/b`, respectively. This would make them look like they have an edit-edit conflict on the sparse directory entry, but if we expand the directory there would not be such a conflict.

File: t/t1092-sparse-checkout-compatibility.sh
Comment (derrickstolee) [inline t/t1092-sparse-checkout-compatibility.sh]: Do you want to try something with `--prefix=deep/deeper2` instead of creating this `test-dir`?

File: unpack-trees.c
Comment (derrickstolee) [inline unpack-trees.c]: "If the prefix is equal to or contained within..." 

My first read of "is or is" was jarring.

File: unpack-trees.c
Comment (derrickstolee) [inline unpack-trees.c:1771]: Oddly, we _want_ the paths to end with a slash, because "A/Dir" is a prefix of "A/Directory" but "A/Dir/" is not a prefix of "A/Directory/". One way to check for this is actually to check that `ce_len == prefix_len || o->prefix[ce_len] == '/'` in the same condition as the `!strcmp()`.

File: t/perf/p2000-sparse-operations.sh
Comment (derrickstolee) [inline t/perf/p2000-sparse-operations.sh:115]: If you reorder this to be early in the series, you can then advertise the results of this test in the commit message for the commit that improves the performance.

**Review by derrickstolee — COMMENTED** [review]
Great start. There are some interesting bits in here, and I really appreciate the fixes for previously-merged code.

I have some recommendations to look into new test cases or tricky situations.

File: t/t1092-sparse-checkout-compatibility.sh
Comment (vdye) [reply t/t1092-sparse-checkout-compatibility.sh]: Ah, I completely missed this case (and you're right - it does cause a conflict). Since the conflict is only detected once you're already in `twoway_merge` or `threeway_merge` (in the middle of `traverse_trees`), you wouldn't be able to expand the index to resolve that conflict. 

I _think_ one way to handle this might be the way `oneway_diff` does (using `diff_tree_oid` to get the differences at a file level, then reconstructing the two- and three-way merges from series of one-way diffs). Since you wrote that addition to `oneway_diff`, I wanted to check - does that make sense? Or were you thinking something different for solving this?

File: read-cache.c
Comment (vdye) [reply read-cache.c:617]: Thanks, that certainly helped clean up some code duplication!

File: wt-status.c
Comment (vdye) [reply wt-status.c]: I'll check this (and add the performance numbers to the commit message) once the functional issues in this series are ironed out. For now, I'll leave this conversation open so I don't forget about it. 🙂

File: t/t1092-sparse-checkout-compatibility.sh
Comment (vdye) [reply t/t1092-sparse-checkout-compatibility.sh]: When `read-tree`-ing with only one tree argument, `bind_merge(...)` prevents "merging" when a file already exists within the prefix. If I run `git read-tree --prefix=deep/deeper2 update-deep`, I get the error:

```
error: Entry 'deep/deeper2/a' overlaps with 'deep/deeper2/a'.  Cannot bind.
```

To avoid creating the new directory, though, I adjusted the test to instead remove the files in `deep/deeper2`, commit that to the test branch, then run `git read-tree --prefix=deep/deeper2 deepest`, which ends up testing the same(ish) scenario.

File: unpack-trees.c
Comment (vdye) [reply unpack-trees.c:1771]: I went with an `if (opposite of that condition) continue;` block (to mirror the style of the rest of the block, but otherwise this should be fixed. Good catch!

File: t/perf/p2000-sparse-operations.sh
Comment (vdye) [reply t/perf/p2000-sparse-operations.sh:115]: I moved the commit, but haven't added performance numbers yet (like with the `git status` change, I'll add those once any outstanding bugs are fixed). In the meantime, I'll keep this open to remind myself.

File: t/t1092-sparse-checkout-compatibility.sh
Comment (derrickstolee) [reply t/t1092-sparse-checkout-compatibility.sh]: I _think_ that's the best way to attempt handling this. The tricky part is constructing the trees again afterwards, but maybe you can plug into the merge machinery to merge the trees to get a new tree? The ORT API might be a way to go here.

The other idea would be to report that the operation failed due to a sparse directory, then expand and try again. This isn't a great idea, but would be simpler to implement.

File: unpack-trees.c
Comment (derrickstolee) [reply unpack-trees.c:1771]: DeMorgan to the rescue!

File: t/t1092-sparse-checkout-compatibility.sh
Comment (vdye) [reply t/t1092-sparse-checkout-compatibility.sh]: I added a _tentative_ implementation for handling two-way merges (which, ultimately, are merging three trees: the index, input_tree_a, and input_tree_b). It passes the tests, but I'm still worried about all of the different edge cases of merges that could happen (e.g., `merge-ort` might successfully merge the contents of a file modified in two branches, but the existing `twoway_merge` might not). And, I haven't added anything yet for `threeway_merge` since the `n` number of base trees (very specific index conditions) don't currently fit nicely with `merge-ort`.

I do still want to avoid excessive index expansion, though - I _think_ one possibility is to refactor `unpack_trees` itself to, rather than using an element-by-element merge function, use `merge-ort` directly (when `unpack_trees_options.merge` is set). It would also require an extension to `merge-ort` to handle multiple base trees in a non-recursive merge. If it's not a terrible idea, I think it's doable - however, it's pretty substantial and probably warrants its own branch/patch series.

In any case, please look over what I have and let me know if you have any suggestions on direction. It's also potentially possible to wait on the two-way and three-way merge cases, since it looks like stash only ever calls `read-tree` with one tree argument.

File: unpack-trees.c
Comment (derrickstolee) [inline unpack-trees.c]: nit: `} else {`

File: unpack-trees.c
Comment (derrickstolee) [inline unpack-trees.c]: Nit: single-line comments will want to be on a single line.

```
/* Find the first conflicted entry, reject the merge for that. */
```

File: merge-ort.c
Comment (derrickstolee) [inline merge-ort.c]: nit: `if (...) {`

It's a hard habit to re-learn.

File: builtin/read-tree.c
Comment (derrickstolee) [inline builtin/read-tree.c:246]: My guess is that this case is rare enough that we might not even bother with it. Let's see if `git stash` ever needs this case, for instance. This `ensure_full_index()` might be a long-term solution.

File: t/t1092-sparse-checkout-compatibility.sh
Comment (derrickstolee) [inline t/t1092-sparse-checkout-compatibility.sh:1257]: This is a nice way to prepare for showing new cases in the future.

File: t/t1092-sparse-checkout-compatibility.sh
Comment (derrickstolee) [inline t/t1092-sparse-checkout-compatibility.sh]: nit: this whitespace doesn't look right (on GitHub, maybe it's right in the editor).

```
	for MERGE_TREES in "update-folder1 update-folder2" \
			   "update-folder2"
```

**Review by derrickstolee — APPROVED** [review]
I like your new approach. I only found style-based nitpicks.

I think leaving the last `ensure_full_index()` is something we can handle for now, unless we find it is critical to the success of the `git stash` work.

File: unpack-trees.c
Comment (vdye) [inline unpack-trees.c]: @derrickstolee After thinking about this more, `merge-ort` isn't _quite_ right for this case (as it is now, anyway) - `merge-ort` performs content merges, whereas `twoway_merge` does not and has its own set of rules for how to "merge" files. Since `stash` only uses the `oneway_merge` case of `read-tree`, I'd like to revert back to d25d0f2e1020e97eea8c51acfef86b9286392113 (with the code style cleanup you've pointed out). That should still get the performance benefit in `stash` (I'll verify that with the tests, though), as well as not introducing functionality that creates mismatches between sparse index and non-sparse index checkouts.

Comment (derrickstolee) [reply unpack-trees.c]: That works for me. Let's not make any unnecessary changes here.

File: t/perf/p2000-sparse-operations.sh
Comment (vdye) [reply t/perf/p2000-sparse-operations.sh:115]: Performance numbers added to commit message - ~90% reduction in execution time!

File: wt-status.c
Comment (vdye) [reply wt-status.c]: It turns out that, even on full-index `git status` the `recursive` flag didn't affect performance in any noticeable way! Probably because a `git status` diff isn't normally running a diff on a directory anyway. Sparse index check removed + performance results added in commit message.

File: builtin/read-tree.c
Comment (vdye) [reply builtin/read-tree.c:246]: Since `git stash` only needs `oneway_diff` (single-tree) cases to work, I'll leave the `ensure_full_index()` calls in for two-way and three-way merges.

**Pull request merged**
